src/compound/tagged_union.ts

```diff
@@ -0,0 +1,73 @@
+import { u8 } from "../primitives/mod.ts";
+import {
+  AlignedType,
+  type InnerType,
+  type Options,
+  type Packed,
+  type ValueOf,
+} from "../types/mod.ts";
+
+type Fn<T> = (value: T) => number;
+
+export class TaggedUnion<
+  T extends Record<number, AlignedType<unknown>>,
+  V extends ValueOf<{ [K in keyof T]: InnerType<T[K]> }> = ValueOf<
+    { [K in keyof T]: InnerType<T[K]> }
+  >,
+> extends AlignedType<V> implements Packed<V> {
+  #record: T;
+  #variantFinder: Fn<V>;
+
+  constructor(input: T, variantFinder: Fn<V>) {
```

Comment:
```suggestion
  constructor(input: T, discriminator: Type<...> = u8) {
```
The discriminator should be a type. That way the discriminator is not an `u8` but anything that can be keyed with.